base/shared/src/main/scala/scalaz/ct/bind.scala

```diff
@@ -3,27 +3,12 @@ package ct
 
 import scala.language.experimental.macros
 
+@meta.minimal("flatMap", "flatten")
 trait BindClass[M[_]] extends ApplyClass[M] {
-  def flatMap[A, B](ma: M[A])(f: A => M[B]): M[B]
-  def flatten[A](ma: M[M[A]]): M[A]
-}
-
-object BindClass {
-
-  trait DeriveAp[M[_]] extends BindClass[M] with MonadClass.Alt[DeriveAp[M]] {
-    final override def ap[A, B](fa: M[A])(f: M[A => B]): M[B] = flatMap(f)(map(fa))
-  }
-
-  trait DeriveFlatten[M[_]] extends BindClass[M] with Alt[DeriveFlatten[M]] {
-    final override def flatten[A](ma: M[M[A]]): M[A] = flatMap(ma)(identity)
-  }
-
-  trait DeriveFlatMap[M[_]] extends BindClass[M] with Alt[DeriveFlatMap[M]] {
-    final override def flatMap[A, B](ma: M[A])(f: (A) => M[B]): M[B] = flatten(map(ma)(f))
-  }
-
-  trait Alt[D <: Alt[D]]
+  def flatMap[A, B](ma: M[A])(f: A => M[B]): M[B] = flatten(map(ma)(f))
```

Comment:
Could this cause an issue with default impl of `map ` in `Monad`? ie if one override only `flatten` then we can have a circular definition between `map` and `flatMap`:
```scala
  def flatMap[A, B](ma: M[A])(f: A => M[B]): M[B] = flatten(map(ma)(f))
  def map[A, B](ma: M[A])(f: A => B): M[B] = flatMap(ma)(a => pure(f(a)))
```
this case was checked by the "`Alt`"-encoding, would be nice to preserve this check.
I think there is some other similar cases as well.